feat: ADR for a new mechanism to automatically log in on the Twig storefront when opened by a logged-in customer via Store-API (#7693)
shopware/platform: adr/2025-03-19-should-have-a-mechanism-to-automatically-login-on-storefront-page-when-opened-it-by-a-logged-in-customer-in-shopware-frontend.md

---

> title: There should be a mechanism to automatically log in on the storefront page when opened by a logged-in customer in the Shopware frontend.

Reviewer: I would not speak of "Shopware frontends" but about the Store-API in general. I guess this should be applicable to every consuming application of the Store-API

Author: Thank you for pointing out. I agree with that 😁

date: 2025-03-19
area: storefront
tags: [storefront, authentication, login, customer, shopware-frontend]
---

## Context
### Current Issue

> - When a customer is already logged into the Shopware frontend and opens the storefront page, they are not automatically logged in. As a result, they have to log in again.

Reviewer: ```suggestion
- If a customer is already logged via Store-API and opens the default Twig storefront, they are not automatically logged in. As a result, they have to log in again.
```

### Use Case Scenario
- In the **Digital Sales Room** plugin, when a customer adds a product to the cart and then opens the checkout page from the storefront page, they are required to log in again.

### Potential Impact

> - This issue could affect other shops using the Shopware frontend, requiring customers to log in multiple times unnecessarily.

Reviewer: ```suggestion
- This issue could affect other shops using the Store-API, requiring customers to log in multiple times unnecessarily.
```

## Decision

> - We will implement a mechanism to automatically log in on the storefront page when opened by a logged-in customer in the Shopware frontend.

Reviewer: ```suggestion
- We will implement a mechanism to automatically log in on the Twig storefront, if opened by a via Store-API logged-in customer.
```

        
        return $this->redirectToRoute($redirectRoute);
    }   
}
```

### Shopware Frontend Implementation
automatically-login.service.ts
```javascript
generateCustomerCsrfToken() {
    const route = 'store-api/account/generated-customer-csrf-token';
    const headers = {
        ...,
        'sw-access-token': 'foo-bar',
    }
    return this.httpClient.post(
        route,
        {},
        { headers },
    );
}

redirectToSalesChannelUrl(salesChannelDomainUrl: string, token: string, customerId: string) {
    const form = document.createElement('form');
    form.method = 'POST';
    form.action = `${salesChannelDomainUrl}/account/login/automatic`;
    form.target = '_blank';
    document.body.appendChild(form);

    this.#createHiddenInput(form, 'token', token);
    this.#createHiddenInput(form, 'customerId', customerId);
    this.#createHiddenInput(form, 'redirectRoute', 'frontend.checkout.confirm.page');

    form.submit();
    form.remove();
}
```

## Consequences
### Security
- The implementation must ensure that the automatic login mechanism is secure when performing the request from other sites by using **CSRF tokens**.

### Benefits

> - The automatic login mechanism will **enhance user experience** by allowing customers who are already logged into the Shopware frontend to access the storefront page without having to log in again.

Reviewer: ```suggestion
- The automatic login mechanism will **enhance user experience** by allowing customers who are already logged in via Store-API to access the Twig storefront without having to log in again.
```

---
title: There should be a mechanism to automatically log in on the Twig storefront when opened by a logged-in customer via Store-API.
date: 2025-03-19
area: storefront
tags: [storefront, checkout, authentication, login, customer, shopware-frontend]
---

## Context
### Current Issue
- If a customer is already logged via Store-API and opens the default Twig storefront, they are not automatically logged in. As a result, they have to log in again.

### Use Case Scenario
- In the **Digital Sales Room** plugin, when a customer adds a product to the cart and then opens the checkout page from the storefront page, they are required to log in again.

### Potential Impact
- This issue could affect other shops using the Store-API, requiring customers to log in multiple times unnecessarily.

## Decision
- We will implement a mechanism to automatically log in on the Twig storefront, if opened by a via Store-API logged-in customer.

### Backend Implementation
#### CustomerCSRFTokenManager
```php
class CustomerCSRFTokenManager
{
    public function generate(string $customerId, string $salesChannelId): string
    {

>         $tokenData = [
>             'salesChannelId' => $salesChannelId,
>             'customerId' => $customerId,
>         ];

Reviewer 2: Is it worth looking at additional fields here for a more secure token?

```suggestion
        $tokenData = [
            'salesChannelId' => $salesChannelId,
            'customerId' => $customerId,
            'expiry' => date('Y-m-d H:i:s', strtotime(' +30 minutes ')),
            'nonce' => 'somebytes',
            'scope' => 'scope', // maybe?
        ];
```

>         return $this->encrypt(hash_hmac(self::HMAC_HASH_ALGORITHM, json_encode($tokenData), $this->appSecret) . '.' . time());

Reviewer 3: Not sure, what this `time()` is doing here. This makes it never verifiable because the token will change every second.
At the same time, the token must expire at some point, otherwise it you can login forever.

Reviewer 2: plus if the timestamp is needed or used (I don't know HMAC encryption that well), wouldn't the unencrypted timestamp then be changeable without invalidating the HMAC? Shouldn't it also be encrypted?

        // Using `hash_equals`
        return true;
    }
}
```

#### GenerateCustomerCSRFTokenRoute
```php
#[Route(defaults: ['_loginRequired' => true])]
public function generateToken(Request $request, CustomerEntity $customer, SalesChannelContext $context): JsonResponse
{
    $csrfToken = $this->customerCsrfTokenManager->generate($customer->getId(), $salesChannelContext->getSalesChannel()->getId());

    return new JsonResponse(['token' => $csrfToken]);
}
```

#### AutomaticLoginRoute
```php

#[Route(path: '/store-api/account/login/automatic', methods: ['POST'])]
public function login(RequestDataBag $requestDataBag, SalesChannelContext $context): ContextTokenResponse
{
    $this->validateRequestDataFields(...);
    $swContextToken = $requestDataBag->getString('swContextToken');
    $session = $this->contextPersister->load($swContextToken, $context->getSalesChannel()->getId());
    $customerId = $session['customerId'];
    $salesChannelId = $session['salesChannelId'];
    if ($customerId && $context->getCustomerId() === $customerId) {
        // return if the customer is already logged in
    }

    $csrfToken = $requestDataBag->getString('csrfToken');

>     $this->customerCsrfTokenManage->validate($csrfToken, $customerId, $salesChannelId);

Reviewer 3: I think, it is dangerous to get the context token from the plain request data. I would say, it should be part of the encrypted token, probably via a JWT token.
